pkg/tsdb/tsdb.go

```diff
@@ -0,0 +1,221 @@
+package tsdb
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/openhistogram/circonusllhist"
+	"go.k6.io/k6/metrics"
+)
+
+type Tag struct {
+	Key, Value string
+}
+
+type TagSet []*Tag
+
+func NewTagSet(ncap int) TagSet {
+	return make([]*Tag, 0, ncap)
+}
+
+// Add adds a new item to the inner slice in an ordered way.
+// Add is a not safe concurrent-operation.
+func (set *TagSet) Add(t *Tag) {
+	index := sort.Search(len(*set), func(i int) bool {
+		return (*set)[i].Key > t.Key
+	})
+
+	if len(*set) == index {
+		*set = append(*set, t)
+		return
+	}
+
+	*set = append((*set)[:index], append(make([]*Tag, 1), (*set)[index:]...)...)
+	(*set)[index] = t
```

Comment:
Isn't this going to be much less efficient than adding all of the tags unsorted and then sorting the whole slice once at the end? This will just do a binary search on every insertion and then copy (potentially big) chunks of the slice every time an element is added... :confused: